Source/Initialization/WarpXInitData.cpp

```diff
@@ -1228,3 +1289,152 @@ void WarpX::CheckKnownIssues()
                 ablastr::warn_manager::WarnPriority::low);
         }
 }
+
+#if defined(WARPX_USE_OPENPMD) && !defined(WARPX_DIM_1D_Z)
+void
+WarpX::ReadExternalFieldFromFile (
+       std::string read_fields_from_path, MultiFab* mf,
+       std::string F_name, std::string F_component)
+{
+    // Get WarpX domain info
+    auto& warpx = WarpX::GetInstance();
+    amrex::Geometry const& geom0 = warpx.Geom(0);
+    const amrex::RealBox& real_box = geom0.ProbDomain();
+    const auto dx = geom0.CellSizeArray();
+    amrex::IntVect nodal_flag = mf->ixType().toIntVect();
+
+    // Read external field openPMD data
+    auto series = openPMD::Series(read_fields_from_path, openPMD::Access::READ_ONLY);
+    auto iseries = series.iterations.begin()->second;
+    auto F = iseries.meshes[F_name];
+    auto offset = F.gridGlobalOffset();
+    auto offset0 = offset[0];
+    auto offset1 = offset[1];
+#if defined(WARPX_DIM_3D)
+    auto offset2 = offset[2];
+#endif
+    auto d = F.gridSpacing<long double>();
+    Real d0 = d[0];
+    Real d1 = d[1];
+#if defined(WARPX_DIM_3D)
+    Real d2 = d[2];
+#endif
+
+    auto FC = F[F_component];
+    auto extent = FC.getExtent();
+
+    // Determine the chunk data that will be loaded.
+    // Now, the full range of data is loaded.
+    // Loading chunk data can speed up the process.
+    // Thus, `chunk_offset` and `chunk_extent` should be modified accordingly in another PR.
+    openPMD::Offset chunk_offset = {0,0,0};
+    openPMD::Extent chunk_extent = {extent[0],extent[1],extent[2]};
+
+    auto FC_chunk_data = FC.loadChunk<double>(chunk_offset,chunk_extent);
+    series.flush();
+    auto FC_data_host = FC_chunk_data.get();
+
+    // Load data to GPU
+    int extent0 = extent[0];
+    int extent1 = extent[1];
+    int extent2 = extent[2];
+    size_t total_extent = size_t(extent[0]) * extent[1] * extent[2];
+    amrex::Gpu::DeviceVector<double> FC_data_gpu(total_extent);
+    auto FC_data = FC_data_gpu.data();
+    amrex::Gpu::copy(amrex::Gpu::hostToDevice, FC_data_host, FC_data_host + total_extent, FC_data);
+
+    // Loop over boxes
+    for (MFIter mfi(*mf, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+        auto box = mfi.growntilebox();
+        const amrex::Box& tb = mfi.tilebox(nodal_flag, mf->nGrowVect());
+        auto const& mffab = mf->array(mfi);
```

Comment:
Same comment on the `auto` types, I guess we could easily replace them with the explicit types, like `amrex::Box` and `amrex::Array4<amrex::Real>` or similar. Whenever possible, I personally find it less criptic and easier for developers to use for debugging, extensions, etc.